### ISSUE TITLE
Current implementation is blocking.

### ISSUE THREAD

**Issue by graysonarts** [RussTheAerialist/node-spi#1]
The current implementation of the native api is blocking.

**Comment by tysseng** [comment]
Hi - I am a bit of a beginner here so this may be obvious to everyone but me, but does this mean that any callbacks on read/transmission are called synchronously, i.e. no other js code will ever be executed between the call to .read(...) and the code inside the callback?

**Comment by Gamadril** [comment]
That's confusing me, too. If write is blocking and actually should not return anything why there is a callback required as second argument?

**Comment by ibash** [comment]
@tysseng yes that's right, because the call to `transfer` here is sync, `callback` is called after the transfer is done and no other js can be executed: https://github.com/RussTheAerialist/node-spi/blob/master/spi.js#L71

@Gamadril AFAICT the second callback is not required, the library checks if it's a function and so if you don't pass anything it will still work: https://github.com/RussTheAerialist/node-spi/blob/master/spi.js#L74

**Comment by Gamadril** [comment]
@ibash you are right it's optional, thanks for the clarification.
I see no disadvantage in beeing blocking. I additionally have to use the sleep module for a millisecond sleep to fullfill the SPI data protocol requirements for the device I'm playing with.
